File: src/algorithms/binary_search.py
def binary_search(structure: list[int], term: int) -> bool:
    low = 0
    high = len(structure)

    while low < high:
        half = low + (high - low) // 2
        value = structure[half]

        if value == term:
            return True

        elif value > term:
            high = half

        else:
            low = half + 1

    return False


def test() -> bool:
    """Throws AssertionError if failed."""

    test_1 = [1, 2, 4, 6, 10, 20, 25, 50, 90]
    assert binary_search(test_1, 1)
    assert binary_search(test_1, 10)
    assert binary_search(test_1, 90)

    assert not binary_search(test_1, 11)

    return True


if __name__ == "__main__":
    try:
        if test():
            print("Test passed!")

    except AssertionError:
        print("Test failed!")
        raise

File: java/src/main/java/algorithms/BinarySearch.java
package algorithms;

import java.util.List;

public class BinarySearch {

    /**
     * A binary search implementation
     *
     * @param list List to be searched
     * @param term The search term
     * @return Returns whether the search term is in the list
     * @param <T> The type generic
     */
    public static <T extends Comparable<T>> boolean binarySearch(List<T> list, T term) {
        int high = list.size();
        int low = 0;

        while (low < high) {
            int half = low + Math.floorDiv(high - low, 2);
            int comparison = term.compareTo(list.get(half));

            if (comparison == 0) {
                return true;
            } else if (comparison < 0) {
                high = half;
            } else {
                low = half + 1;
            }
        }
        return false;
    }
}
